src/extract_gpuinfo_msm.c

```diff
@@ -0,0 +1,523 @@
+      }
+    }
+
+    if (fd < 0)
+      continue;
+
+    drmVersionPtr ver = _drmGetVersion(fd);
+
+    if (!ver) {
+      close(fd);
+      continue;
+    }
+
+    bool is_msm = !strcmp(ver->name, "msm");
+
+    if (!is_msm) {
+      _drmFreeVersion(ver);
+      close(fd);
+      continue;
+    }
+
+    authenticate_drm(fd);
+
+    gpu_infos[msm_gpu_count].drmVersion = ver;
+    gpu_infos[msm_gpu_count].fd = fd;
+    gpu_infos[msm_gpu_count].base.vendor = &gpu_vendor_msm;
+
+    list_add_tail(&gpu_infos[msm_gpu_count].base.list, devices);
+    // Register a fdinfo callback for this GPU
+    processinfo_register_fdinfo_callback(parse_drm_fdinfo_msm, &gpu_infos[msm_gpu_count].base);
+    msm_gpu_count++;
+  }
+
+  _drmFreeDevices(devs, libdrm_count);
+  *count = msm_gpu_count;
+
+  return true;
+}
+
+static int gpuinfo_msm_query_param(int gpu, uint32_t param, uint64_t *value) {
+  struct drm_msm_param req = {
+    .pipe = MSM_PIPE_3D0, // Only the 3D pipe.
+    .param = param,
+  };
+
+  int ret = _drmCommandWriteRead(gpu, DRM_MSM_GET_PARAM, &req, sizeof(req));
+
+  if (ret)
+    return ret;
+
+  *value = req.value;
+
+  return 0;
+}
+
+void gpuinfo_msm_populate_static_info(struct gpu_info *_gpu_info) {
+  struct gpu_info_msm *gpu_info = container_of(_gpu_info, struct gpu_info_msm, base);
+  struct gpuinfo_static_info *static_info = &gpu_info->base.static_info;
+  const char *dev_name;
+
+  static_info->integrated_graphics = true;
+  RESET_ALL(static_info->valid);
+
+  uint64_t gpuid;
+  if (gpuinfo_msm_query_param(gpu_info->fd, MSM_PARAM_CHIP_ID, &gpuid) == 0) {
+    const char* name = msm_parse_marketing_name(gpuid);
+    if (name) {
+      strncpy(static_info->device_name, name, sizeof(static_info->device_name));
+    }
+    else {
+      snprintf(static_info->device_name, sizeof(static_info->device_name), "Unknown Adreno %lx", gpuid);
+    }
+    SET_VALID(gpuinfo_device_name_valid, static_info->valid);
+  }
+}
+
+void gpuinfo_msm_refresh_dynamic_info(struct gpu_info *_gpu_info) {
+  struct gpu_info_msm *gpu_info = container_of(_gpu_info, struct gpu_info_msm, base);
+  struct gpuinfo_static_info *static_info = &gpu_info->base.static_info;
+  struct gpuinfo_dynamic_info *dynamic_info = &gpu_info->base.dynamic_info;
+
+  RESET_ALL(dynamic_info->valid);
+  dynamic_info->encode_decode_shared = true;
+
+  // GPU clock
+  uint64_t val;
+  if (gpuinfo_msm_query_param(gpu_info->fd, MSM_PARAM_MAX_FREQ, &val) == 0) {
+    // TODO: No way to query current clock speed.
+    SET_GPUINFO_DYNAMIC(dynamic_info, gpu_clock_speed, val / 1000000);
+    SET_GPUINFO_DYNAMIC(dynamic_info, gpu_clock_speed_max, val / 1000000);
+  }
+
+  // Mem clock
+  // TODO: No way to query.
+
+  // TODO: find how to extract global utilization
+  // gpu util will be computed as the sum of all the processes utilization for now
+
+  struct sysinfo info;
+  if (sysinfo(&info) == 0) {
+    SET_GPUINFO_DYNAMIC(dynamic_info, total_memory, info.totalram);
```

Comment:
Has the GPU access to the entirety of the RAM or is this an approximation to get the info to show up?

Reply:
yes, GPU can access all RAM.. or rather any RAM can be mapped to the GPU.  You might want to use `MSM_PARAM_VA_SIZE` to get the size of the GPU's address space, which might be less than "all RAM" (ie. on some devices it is 4GB.. but even in this case each process using the GPU has it's own 4GB address space so a combination of multiple processes could map all RAM).